### PR TITLE
feat: make the languageId of didOpen notifications configurable

### PR DESCRIPTION
**Pull request by ayame113**
PR for https://github.com/atom-community/atom-languageclient/issues/178.

### CODE REVIEW

File: lib/adapters/document-sync-adapter.ts [hunk 1]
```diff
@@ -149,16 +153,17 @@ export class TextEditorSyncAdapter {
...
-   * @param documentSync The document syncing options.
+   * @param _editor A {TextEditor} to keep in sync.
+   * @param _connection A {LanguageClientConnection} to a language server to keep in sync.
+   * @param _documentSync The document syncing options.
    */
   constructor(
     private _editor: TextEditor,
     private _connection: LanguageClientConnection,
     private _documentSync: TextDocumentSyncOptions,
     private _versions: Map<string, number>,
```
Comment (ayame113) [inline lib/adapters/document-sync-adapter.ts:164]: There are minor changes to enable JSDoc that are unrelated to the issue.

**Review by UziTech — APPROVED** [review]
Nice work! :100:

File: lib/auto-languageclient.ts [hunk 2]
```diff
@@ -295,6 +295,25 @@ export default class AutoLanguageClient {
...
+   *         if (editor.getGrammar().scopeName === "source.myLanguage") {
+   *           return "myCustumLanguageId"
+   *         }
+   *       }
+   *     }
+   *
+   * @param _editor A {TextEditor} which is opened.
+   * @returns A {string} of `languageId` used for `textDocument/didOpen` notification.
+   */
+  public getLanguageIdFromEditor(_editor: TextEditor): string | void | null {}
```
Comment (aminya) [inline lib/auto-languageclient.ts:315]: I would prefer to have a working function and use it as it is. The client can override a custom behaviour if they want to.

```suggestion
  protected getLanguageIdFromEditor(_editor: TextEditor): string | void | null {
    return this._editor.getGrammar().name
  }
```

**Review by aminya — CHANGES_REQUESTED** [review]

File: lib/adapters/document-sync-adapter.ts [hunk 3]
```diff
@@ -203,9 +208,12 @@ export class TextEditorSyncAdapter {
...
   }
 
-  /** Get the languageId field that will be sent to the language server by simply using the grammar name. */
+  /**
+   * Get the languageId field that will be sent to the language server by simply using the `_getLanguageIdFromEditor`.
+   * Fall back to the grammar name if `_getLanguageIdFromEditor` returns null or undefined.
+   */
   public getLanguageId(): string {
-    return this._editor.getGrammar().name
+    return this._getLanguageIdFromEditor(this._editor) ?? this._editor.getGrammar().name
```
Comment (aminya) [inline lib/adapters/document-sync-adapter.ts:216]:
```suggestion
    return this._getLanguageIdFromEditor(this._editor)
```

**Comment by ayame113** [comment]
Thank you for your review!

@aminya 
I fixed it as mentioned in the comment.

**Review by aminya — APPROVED** [review]
Thanks for your contribution!

**Pull request merged**